core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -501,11 +505,13 @@ class KafkaApis(val requestChannel: RequestChannel,
         } else {
           (metadataCache.getTopicName(topic.topicId).orElse(topic.name), topic.topicId())
         }
-
         val topicPartition = new TopicPartition(topicName, partition.index())
+
         // To be compatible with the old version, only return UNKNOWN_TOPIC_ID if request version uses topicId, but the corresponding topic name can't be found.
         if (topicName.isEmpty && request.header.apiVersion > 12)
           nonExistingTopicResponses += new TopicIdPartition(topicId, topicPartition) -> new PartitionResponse(Errors.UNKNOWN_TOPIC_ID)
+        else if (!metadataCache.asInstanceOf[KRaftMetadataCache].getImage().topics().getTopic(topic.topicId()).partitions().get(partition.index()).clusterLinkName.isEmpty())
+          readOnlyTopicResponses += new TopicIdPartition(topicId, topicPartition) -> new PartitionResponse(Errors.INVALID_REQUEST)
```

Comment:
There is a bug, we should also check `metadataCache.asInstanceOf[KRaftMetadataCache].getImage().topics().getTopic(topic.topicId()) != null`